src/NUnitFramework/framework/Internal/Builders/TestCache.cs

```diff
@@ -0,0 +1,92 @@
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+using NUnit.Compatibility;
+
+namespace NUnit.Framework.Internal.Builders
+{
+    /// <summary>
+    /// Caches static information for ITestAction to reduce re-calculations and memory allocations from reflection.
+    /// </summary>
+    internal static class TestCache
+    {
+#if NET35
+        private static readonly Dictionary<Type, TestMetadata> TestMetadataCache = new Dictionary<Type, TestMetadata>();
+#else
+        private static readonly ConcurrentDictionary<Type, TestMetadata> TestMetadataCache = new ConcurrentDictionary<Type, TestMetadata>();
+#endif
+
+        /// <summary>
+        /// Returns cached metadata for method instance.
+        /// </summary>
+        internal static TestMetadata Get(Type testType)
+        {
+#if NET35
+            lock (TestMetadataCache)
+            {
+                if (!TestMetadataCache.TryGetValue(testType, out var metadata))
+                    TestMetadataCache.Add(testType, metadata = new TestMetadata(testType));
+
+                return metadata;
+            }
+#else
+            return TestMetadataCache.GetOrAdd(testType, m => new TestMetadata(m));
+#endif
+        }
+
+        /// <summary>
+        /// Memoization of Test information to reduce subsequent allocations from parameter and attribute information.
+        /// </summary>
+        internal sealed class TestMetadata
+        {
+            public TestMetadata(Type testType)
+            {
+                TestActionAttributes = GetActionsForType(testType);
+            }
+
+            public ITestAction[] TestActionAttributes { get; }
+
+            private static ITestAction[] GetActionsForType(Type type)
+            {
+                var actions = new List<ITestAction>();
+
+                if (type != null && type != typeof(object))
```

Comment:
can we waely exist here and do return an array.empty

Reply:
Done